src/aura/CRLP_FilterGroup/CRLP_FilterGroupHelper.js

```diff
@@ -109,35 +286,107 @@
+        cmp.set("v.notificationText", text);
+        cmp.set("v.notificationClasses", "");
+    },
+
+    /**
+     * @description: toggles a modal popup and backdrop
+     */
+    toggleFilterRuleModal: function(cmp) {
         var backdrop = cmp.find('backdrop');
         $A.util.toggleClass(backdrop, 'slds-backdrop_open');
         var modal = cmp.find('modaldialog');
         $A.util.toggleClass(modal, 'slds-fade-in-open');
     },
 
-    /* @description: verifies all required fields have been populated before saving the filter group
+    /**
+     * @description Show or Hide the page spinner
+     * @param showSpinner - true to show; false to hide
      */
-    validateFilterGroupFields: function(cmp){
+    toggleSpinner: function(cmp, showSpinner) {
+        var spinner = cmp.find("waitingSpinner");
+        if (showSpinner === true) {
+            $A.util.removeClass(spinner, "slds-hide");
+        } else {
+            $A.util.addClass(spinner, "slds-hide");
+        }
+    },
+
+    /**
+     * @description: verifies all required fields have been populated before saving the filter group
+     * @return: boolean canSave confirming if filter group can be saved
+     */
+    validateFilterGroupFields: function(cmp) {
         var canSave = true;
         var name = cmp.get("v.activeFilterGroup.MasterLabel");
         var description = cmp.get("v.activeFilterGroup.Description__c");
         var filterRuleList = cmp.get("v.filterRuleList");
 
         cmp.find("nameInput").showHelpMessageIfInvalid();
         cmp.find("descriptionInput").showHelpMessageIfInvalid();
-        //todo: where to put error message for filter rules? or can a filter group be saved w/o filter rules?
 
-        if(!description){
+        if (!description) {
             canSave = false;
-        } else if(!name){
+        } else if (!name) {
             canSave = false;
-        } else if(filterRuleList.length === 0){
+        } else if (filterRuleList.length === 0) {
             canSave = false;
         }
 
         return canSave;
+    },
+
+    /**
+     * @description: verifies filter rule that all fields are provided and isn't a duplicate
+     * @return: boolean canSave confirming if filter group can be saved
+     */
+    validateFilterRuleFields: function(cmp, filterRule, filterRuleList) {
+
+        //check for field validity
+        var canSave = cmp.find("filterRuleField").reduce(function (validSoFar, filterRuleCmp) {
+            filterRuleCmp.showHelpMessageIfInvalid();
+            return validSoFar && filterRuleCmp.get("v.validity").valid;
+        }, true);
+
+        //check for duplicates
+        if (cmp.get("v.filterRuleMode") === 'create') {
+            //check for specific API names instead of the full row to avoid issues with the ID not matching
+            for (var i = 0; i < filterRuleList.length; i++) {
+                if (filterRuleList[i].objectName === filterRule.objectName
```

Comment:
I think these two blocks can be combined into a single for loop. Within the loop, it can check for either `cmp.get("v.filterRuleMode") === 'create'` or `i !== filterRule.index` to determine if the comparisons should be made.